src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs

```diff
@@ -255,25 +255,83 @@ protected override Expression VisitConstant(ConstantExpression c)
-            if (typeArgs != null)
-                return mi.MakeGenericMethod(typeArgs);
-            return mi;
+
+            return matchingMethods[0];
 
             [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
                 Justification = "This method is intentionally hiding the Enumerable type from the trimmer so it doesn't preserve all Enumerable's methods. " +
                 "This is safe because all Queryable methods have a DynamicDependency to the corresponding Enumerable method.")]
             static MethodInfo[] GetEnumerableStaticMethods(Type type) =>
                 type.GetMethods(BindingFlags.Public | BindingFlags.Static);
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod",
+                Justification = "Enumerable methods don't have trim annotations.")]
+            MethodInfo ApplyTypeArgs(MethodInfo methodInfo) => typeArgs == null ? methodInfo : methodInfo.MakeGenericMethod(typeArgs);
+
+            // In certain cases, there might be ambiguities when resolving matching overloads, for example between
+            //   1. FirstOrDefault<object>(IEnumerable<object> source, Func<object, bool> predicate) and
+            //   2. FirstOrDefault<object>(IEnumerable<object> source, object defaultvalue).
+            // In such cases we disambiguate by picking a method with the most derived signature.
+            static MethodInfo DisambiguateMatches(MethodInfo[] matchingMethods)
+            {
+                Debug.Assert(matchingMethods.Length > 1);
+                ParameterInfo[][] parameters = matchingMethods.Select(m => m.GetParameters()).ToArray();
+
+                // `IsLessDerivedThan` defines a partial order on method signatures; pick a maximal element using that order.
+                // It is assumed that `matchingMethods` is a small array, so a naive quadratic search is probably better than
+                // doing some variant of topological sorting.
+
+                for (int i = 0; i < matchingMethods.Length; i++)
+                {
+                    bool foundDerivedMethodSignature = false;
+                    for (int j = 0; j < matchingMethods.Length; j++)
+                    {
+                        if (i != j && IsLessDerivedThan(parameters[i], parameters[j]))
+                        {
+                            foundDerivedMethodSignature = true;
+                            break;
+                        }
+                    }
+
+                    if (!foundDerivedMethodSignature)
+                    {
+                        // Found a maximal element
+                        return matchingMethods[i];
+                    }
+                }
+
+                Debug.Fail("Search should have found a maximal element");
+                throw new Exception();
+
+                static bool IsLessDerivedThan(ParameterInfo[] params1, ParameterInfo[] params2)
```

Comment:
I'm not coming up with any great suggestions, but is there a better name than `IsLessDerivedThan`? What exactly does "less derived" mean? It is also a bit confusing that in the comments it says "by picking a method with the most derived signature." It is also a bit confusing on which parameter is the "left" and which is the "right" in the "less derived" operation.

`AllParametersDerivedFrom`
`AllParameterTypesAssignableFrom`

Reply:
> It is also a bit confusing on which parameter is the "left" and which is the "right" in the "less derived" operation.

I guess I was using convention from binary relations i.e. "IsLessThan(left,right)" meaning "left IsLessThan right", but I agree this might not be too evident in this context. Perhaps calling the parameters "left" and "right" might help self-document this?

Reply:
> Perhaps calling the parameters "left" and "right" might help self-document this?

Yeah, I think that makes it better.